sklearn/utils/estimator_checks.py

```diff
@@ -847,6 +858,139 @@ def _generate_sparse_matrix(X_csr):
         yield sparse_format + "_64", X
 
 
+def check_array_api_input(name, estimator_orig, *, array_namespace):
+    """Check that the array_api Array gives the same results as ndarrays."""
+    xp = pytest.importorskip(array_namespace)
+
+    X, y = make_classification(random_state=42)
+
+    est = clone(estimator_orig)
+
+    X_xp = xp.asarray(X)
+    y_xp = xp.asarray(y)
+
+    est.fit(X, y)
+
+    array_attributes = {
+        key: value for key, value in vars(est).items() if isinstance(value, np.ndarray)
+    }
+
+    est_xp = clone(est)
+    with config_context(array_api_dispatch=True):
+        est_xp.fit(X_xp, y_xp)
+
+    # Fitted attributes which are arrays must have the same
+    # namespace as the one of the training data.
+    for key, attribute in array_attributes.items():
+        est_xp_param = getattr(est_xp, key)
+        assert hasattr(est_xp_param, "__array_namespace__")
+
+        est_xp_param_np = _convert_to_numpy(est_xp_param, xp=xp)
+        assert_allclose(
+            attribute, est_xp_param_np, err_msg=f"{key} not the same", atol=1e-3
```

Comment:
I'm actually a bit surprised to see `atol=1e-3`. Is there a particular reason why we have such a large tolerance? I know it is inherited from the existing tests, but I don't know the reasoning for it originally either. It just seems especially large compared to our tolerances on many other algorithms and implementations.

Reply:
This is comparing the output of results from a GPU and CPU, which can result in bigger differences. For example, PyTorch sets `atol=1e-3` and `rtol=1e-3` when comparing CPU and GPU operations:

https://github.com/pytorch/pytorch/blob/c618093681bfbb7fca6159de5b73ffceb5b4e0f5/test/test_ops.py#L287

Reply:
After seeing Olivier's change to `100 * finfo(dtype).eps`, should we still add a comment?

Reply:
It wouldn't hurt to leave a comment about how these error bounds are tighter than PyTorch which is a strong precedence.